### ISSUE TITLE
Variable explorer could show the last command that updated the variable

### ISSUE THREAD

**Issue by e346615** [Microsoft/RTVS#2980]
In R Interactive, it would be useful if I could right-click on a variable name in the Variable Explorer and see the last command that populated that variable. For example, if I execute the command
**x <- filter(d, rowSums(d) >1)**
then the variable **x** appears in the Explorer window. I would like to be able to get the
**filter(d, rowSums(d) >1)** command back by right-clicking on the **x**.

**Comment by smortaz** [comment]
+1

**Comment by int19h** [comment]
Implementing this in a general fashion is impossible - R doesn't offer any mechanisms that allow for fine grained change tracking on variable and expression level. If you source() a file, for example, we have no way to find out which line in that file did the change.

The most I can see working here is tracking the last REPL input that caused the change, and then only for variables that are already visible in Variable Explorer (i.e. non-expanded children wouldn't be tracked). The reason is that the only way for us to detect a change is to do a full diff between "before" and "after" states. Doing this just for variables that are visible in VE is feasible, but doing it for all their children recursively would basically mean diffing the entire R world - way too slow.

**Comment by smortaz** [comment]
Ack. I think the guesstimate 70% approach is good enough. I.e. ---

The most I can see working here is tracking the last REPL input that caused
the change, and then only for variables that are already visible in
Variable Explorer

**Comment by e346615** [comment]
I agree. I thought it would be available at the beginning of the parsing operation, when the <- or -> was detected.

**Comment by Hong-Revo** [comment]
`<-` and `->` are just the most common ways of modifying a variable. Consider also:

    a = 2
    a <<- 2
    assign("a", 2)
    eval(parse(text="a <- 2"))
    source("file.R")  # contains 'a <- 2'

This doesn't take into account evaluation frames either. For example:

    a <- 2
    local(a <- 3)
    a  # still contains 2

And then you have unevaluated expressions:

    a <- 2
    quote(a <- 3)
    a  # still contains 2

Doing this properly seems like it would be a lot of work and require digging into the guts of R. I'm also not sure what the benefit would be over simply looking at the R history window.

**Comment by MikhailArkhipov** [comment]
@e346615 - R engine parses the expression during the execution so we don't have direct access there. As @int19h said, we can try and track expressions in REPL, but it one sourced the entire file and somewhere in a middle there was an assignment, we won't know it it happened.

**Comment by e346615** [comment]
Agreed, withdrawn.
